articles/what-is-an-abi/index.md

```diff
@@ -0,0 +1,268 @@
+
+We will talk more about these type-kinds in the [structs](#structs) and [calling conventions](#passing-data-to-the-callee-and-back) sections.
+
+When it comes to primitive types the main issues to think about are how to store them in memory and how to operate on them. These two things play hand in hand because if we want to operate on some data we will first need to read them from memory and into a register.
+
+For anything stored in memory, we will have to think about the [endianness](https://www.section.io/engineering-education/what-is-little-endian-and-big-endian/) of the data. Most of the time this is dealt with by the hardware and exposed to the program as a `mov` instructions that already account for endianness with exception to loading constant value with `mov`. I won't go into detail as I wrote a whole article about it [here](https://www.section.io/engineering-education/what-is-little-endian-and-big-endian/).
+
+## Size and Alignment
+
+There are also the size and alignment of each piece of data that we have to worry about. The size is, well the size but must also be a multiple of the alignment. The alignment specifies where it can be stored in memory, it must be at least 1 and always a power of two. As an example, if something has an alignment of 4-bytes then the address it is stored at must be divisible by 4-bytes. The reason why this exists is mostly for [performance reasons](https://stackoverflow.com/q/49391001/9664285) because memory hardware is specialized in such a way to read data from cache-lines and alignment can guarantee that one piece of data is not split up into multiple cache-lines.
+
+Take for example a 4-byte value with a 4-byte alignment and an 8-byte value with an 8-byte alignment next to each other. To store these two values we would have to have 4 bytes of empty space between them (shown by the `x`s).
+
+```
+addr:  0123456789abcdef
+value: aaaaxxxxbbbbbbbb
+```
+
+Most ABIs require the alignment of primitive types to be the same as the size ([table1](https://software.intel.com/sites/default/files/article/402129/mpx-linux64-abi.pdf#figure.3.1), [table2](https://gankra.github.io/blah/rust-layouts-and-abis/#the-layoutsabis-of-builtins)).
+
+# Structs
+
+Like alignment, structs have a similar concept for the data in them. This is to say that structs also have their own alignment so that the data in them can have the alignment they require. We can sort of think about each struct as containing their own space of memory where everything in them is aligned relative to the start of the struct. We call this struct relative memory address the offset of a field in the struct. Let's look at an example.

```

Comment:
We can sort of think about ->  We can think about

Reply:
Fix: We call this struct relative memory address the offset of a field in the struct.